### PR TITLE
Adding Kpneumoniae workflow

### PR DESCRIPTION
**Pull request by ryanjameskennedy**
# Description
_Summary of the changes made:_
* Updated the config
* Added kpneumoniae workflow

## Primary function of PR
- [x] Major functionality improvement / New type of analysis

# Testing
Has not been tested yet. Perhaps @samuell can run a test before merging?

### CODE REVIEW

File: workflows/klebsiella_pneumoniae.nf [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+include { freebayes                                 } from '../nextflow-modules/modules/freebayes/main'
+include { kraken                                    } from '../nextflow-modules/modules/kraken/main'
+include { mask_polymorph_assembly                   } from '../nextflow-modules/modules/mask/main'
+include { mlst                                      } from '../nextflow-modules/modules/mlst/main'
+include { resfinder                                 } from '../nextflow-modules/modules/resfinder/main'
+include { samtools_index as samtools_index_assembly } from '../nextflow-modules/modules/samtools/main'
+include { virulencefinder                           } from '../nextflow-modules/modules/virulencefinder/main'
+include { CALL_BACTERIAL_BASE                       } from '../workflows/bacterial_base.nf'
+
+workflow CALL_STAPHYLOCOCCUS_AUREUS {
```
Comment (samuell) [inline workflows/klebsiella_pneumoniae.nf:24]:
```diff
-workflow CALL_STAPHYLOCOCCUS_AUREUS {
+workflow CALL_KLEBSIELLA_PNEUMONIAE {
```
:)

**Comment by samuell** [comment]
Testing with ...

```bash
nextflow run /home/samuel.lampa/opt/jasen/main.nf \
    -profile klebsiella_pneumoniae \
    -config /home/proj/development/microbial/bacterial_typing/jasen_testruns/nextflow.testrun.config \
    --csv /home/proj/development/microbial/bacterial_typing/jasen_testruns/jasen_inputspec.csv
```

... generates some complaints:

```
N E X T F L O W  ~  version 23.04.1
Launching `/home/samuel.lampa/opt/jasen/main.nf` [deadly_moriondo] DSL2 - revision: 451e961720
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:assembly_trim_clean    -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:save_analysis_metadata -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:skesa                  -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:spades_illumina        -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:spades_iontorrent      -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:quast                  -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:bwa_mem_ref            -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:samtools_index_ref     -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:post_align_qc          -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:CALL_BACTERIAL_BASE:sourmash               -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:bwa_index                                  -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:bwa_mem_dedup                              -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:samtools_index_assembly                    -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:freebayes                                  -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:mask_polymorph_assembly                    -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:mlst                                       -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:chewbbaca_create_batch_list                -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:chewbbaca_allelecall                       -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:chewbbaca_split_results                    -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:amrfinderplus                              -
[-        ] process > CALL_KLEBSIELLA_PNEUMONIAE:resfinder                                  -
WARN: Access to undefined parameter `useVirulenceDbs` -- Initialise it to a default value eg. `params.useVirulenceDbs = some_value`
ERROR ~ A process input channel evaluates to null -- Invalid declaration `val databases`

 -- Check script '/home/samuel.lampa/opt/jasen/./workflows/klebsiella_pneumoniae.nf' at line: 105 or see '.nextflow.log' file for more details
```

(Investigating...)

**Comment by samuell** [comment]
> ... generates some complaints:
> (Investigating...)

Well, this particular one seem to be solved by just adding an empty list in the klebsiella params section in the config:

```
@@ -50,6 +51,7 @@ profiles {
                params.chewbbacaDb = "${params.root}/assets/cgmlst/klebsiella_pneumoniae/alleles_rereffed"
                params.coreLociBed = "${params.root}/assets/cgmlst/klebsiella_pneumoniae/bed/NC_016845.1.bed"
                params.trainingFile = "${params.root}/assets/cgmlst/klebsiella_pneumoniae/alleles_rereffed/Klebsiella_pneumoniae.trn"
+               params.useVirulenceDbs = []
        }
```

**Comment by samuell** [comment]
Now got:

```
Command error:
  [11:09:16] This is mlst 2.23.0 running on linux with Perl 5.032001
  [11:09:16] Checking mlst dependencies:
  [11:09:16] Found 'blastn' => /usr/local/bin/blastn
  [11:09:16] Found 'any2fasta' => /usr/local/bin/any2fasta
  [11:09:17] Found blastn: 2.13.0+ (002013)
  [11:09:17] Invalid --scheme 'kpneumoniae'. Check using --list
```

**Comment by ryanjameskennedy** [comment]
> Now got:
> 
> ```
> Command error:
>   [11:09:16] This is mlst 2.23.0 running on linux with Perl 5.032001
>   [11:09:16] Checking mlst dependencies:
>   [11:09:16] Found 'blastn' => /usr/local/bin/blastn
>   [11:09:16] Found 'any2fasta' => /usr/local/bin/any2fasta
>   [11:09:17] Found blastn: 2.13.0+ (002013)
>   [11:09:17] Invalid --scheme 'kpneumoniae'. Check using --list
> ```

I ran `mlst --list`:
We need to change `kpneumoniae` to `klebsiella`.

**Comment by samuell** [comment]
> I ran mlst --list:

> We need to change kpneumoniae to klebsiella.

I'll make a try!

**Comment by samuell** [comment]
(Confirming that I managed to run through the pipeline fine now).

Some caveats for myself or anyone else:

1. I had to set `runOptions = '--bind /tmp'` in the singularity section in the config file to be able to have Nextflow to use /tmp as the scratch folder.
2. Not completely sure about the minimal settings for the TMP folders, but the following did the trick for me:

  ```bash
  export TMPDIR=/tmp && \
  export TEMPDIR=/tmp && \
  export NXF_TEMP=/tmp && \
  export NXF_SINGULARITY_TMPDIR=/tmp && \
  export SINGULARITY_TMPDIR=/tmp && \
  SINGULARITY_TMPDIR=/tmp && \
  nextflow run /path/to/jasen/main.nf \
    -profile klebsiella_pneumoniae \
    -config /path/to/jasen/jasen_testruns/nextflow.testrun.config \
    -w /tmp/nfwork \
    --csv /path/to/jasen/jasen_testruns/jasen_inputspec.csv
  ```
3. Kraken is a bit picky. It wants either gzipped or not gzipped fastq files :) (default is gzipped, as per a setting in the bottom of the default nextflow config file)

**Pull request merged**